util/src/trie/mod.rs

```diff
@@ -76,6 +77,46 @@ pub type Result<T> = ::std::result::Result<T, Box<TrieError>>;
 /// Trie-Item type.
 pub type TrieItem<'a> = Result<(Vec<u8>, DBValue)>;
 
+/// Description of what kind of query will be made to the trie.
+///
+/// This is implemented for any &mut recorder (where the query will return
+/// a DBValue), any function taking raw bytes (where no recording will be made),
+/// or any tuple of (&mut Recorder, FnOnce(&[u8]))
+pub trait Query {
+	/// Output item.
+	type Item;
+
+	/// Decode a byte-slice into the desired item.
+	fn decode(self, &[u8]) -> Self::Item;
+
+	/// Record that a node has been passed through.
+	fn record(&mut self, &H256, &[u8], u32) { }
+}
+
+impl<'a> Query for &'a mut Recorder {
+	type Item = DBValue;
```

Comment:
Since recording queries often don't care about the value itself it might make sense just to use `()` here.